Adds an "identity" reprojection test (which currently fails)

barentsen opened the pull request:
I have added a unit test which, although pathological, I believe should pass.  At present this unit test fails, and it may reveal a bug (which I do not understand at present).

Apologies for the brevity of this issue for now. I wanted to make that the unconference session at #pyastro15 on this package produced a useful PR! :-)

astrofrog commented:
@barentsen - thanks! I agree that this should definitely pass. Would you be interested in working on investigating why it isn't working? I think a good place to start would be to diagnose why the nearest neighbor case isn't working.

astrofrog commented:
@barentsen - could you rebase this on master to fix some of the unrelated failures?

astrofrog commented:
@barentsen - can you rebase this? I'd like to see if it's still an issue. Thanks!

barentsen commented:
@astrofrog Rebase done!

astrofrog commented:
@barentsen - thanks! The ``reproject`` function no longer exists, need to specify e.g. ``reproject_interp``

astrofrog commented:
@barentsen - could you rebase and also update the function name as mentioned above?

barentsen commented:
@astrofrog I have rebased and changed the test to use `reproject_interp` (and `reproject_exact` for flux-conversing mode).  The test still fails however, which I believe to expose a true bug. 

```Python
$ py.test reproject/tests/test_high_level.py
reproject/tests/test_high_level.py .......................................................................................................................F

================================================================================= FAILURES ==================================================================================
_________________________________________________________________ test_identity_projection[flux-conserving] _________________________________________________________________

projection_type = 'flux-conserving'

    @pytest.mark.parametrize('projection_type', ALL_MODES)
    def test_identity_projection(projection_type):
        header_in = fits.Header.fromstring(IDENTITY_TEST_HDR, sep='\n')
        data_in = np.ones((header_in['NAXIS2'], header_in['NAXIS1']))
        if projection_type == 'flux-conserving':
            data_out, footprint = reproject_exact((data_in, header_in), header_in)
        else:
            data_out, footprint = reproject_interp((data_in, header_in), header_in,
                                                   order=projection_type)
        # When reprojecting with an identical input and output header,
        # we may expect the input and output data to be similar,
        # and the footprint values to be ~ones.
>       np.testing.assert_allclose(footprint, data_in, rtol=0.1)

reproject/tests/test_high_level.py:157: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

x_id = array([[ True, False,  True,  True,  True],
       [ True,  True,  True, False,  True],
       [False,  True,  True,  ...se, False,  True],
       [ True,  True, False, False,  True],
       [False, False,  True, False,  True]], dtype=bool)
y_id = array([[False, False, False, False, False],
       [False, False, False, False, False],
       [False, False, False, F...se, False, False],
       [False, False, False, False, False],
       [False, False, False, False, False]], dtype=bool)
hasval = 'nan'

    def chk_same_position(x_id, y_id, hasval='nan'):
        """Handling nan/inf: check that x and y have the nan/inf at the same
            locations."""
        try:
            assert_array_equal(x_id, y_id)
        except AssertionError:
            msg = build_err_msg([x, y],
                                err_msg + '\nx and y %s location mismatch:'
                                % (hasval), verbose=verbose, header=header,
                                names=('x', 'y'), precision=precision)
>           raise AssertionError(msg)
E           AssertionError: 
E           Not equal to tolerance rtol=0.1, atol=0
E           
E           x and y nan location mismatch:
E            x: array([[ nan,   1.,  nan,  nan,  nan],
E                  [ nan,  nan,  nan,   1.,  nan],
E                  [  3.,  nan,  nan,  nan,  nan],...
E            y: array([[ 1.,  1.,  1.,  1.,  1.],
E                  [ 1.,  1.,  1.,  1.,  1.],
E                  [ 1.,  1.,  1.,  1.,  1.],...

../../bin/anaconda/lib/python3.4/site-packages/numpy/testing/utils.py:656: AssertionError
=================================================================== 1 failed, 119 passed in 4.01 seconds ====================================================================
```

astrofrog commented:
@barentsen - I know it's been a while, but could you rebase? I think this should finally be fixed 🤞

barentsen commented:
@astrofrog No problem!  I rebased and found the test to be passing on my local machine. :tada: :tada: :tada: 

Fingers crossed Travis agrees!
